dfm/cleaning/quality.py

```diff
@@ -0,0 +1,279 @@
+    L. A., Rauh, M., Huang, P.-S., … Irving, G. (2021).
+    Scaling Language Models: Methods, Analysis & Insights from Training Gopher.
+    https://arxiv.org/abs/2112.11446v2
+"""
+
+from collections import Counter
+from email.generator import Generator
+from functools import partial
+from typing import Iterable, Optional, Set, Tuple
+
+import spacy
+from spacy.tokens import Doc
+
+
+class QualityFilter:
+    """
+    Danish implementation of quality filter described in (Rae et al., 2021).
+
+    Args:
+        stop_words (Optional[Set[str]], optional): A set of stop words to use.
+            Defaults to None.
+        min_stop_words (int, optional): The least ammount of stop words a text
+            should have before it is kept. Defaults to 2.
+        mean_word_length (Tuple[int, int], optional): Upper and lower bound on the
+            mean word length. Defaults to (3, 10).
+        doc_length (Tuple[int, int], optional): Upper and lower bound on the
+            documents length. Defaults to [50, 100_000].
+        alpha_ratio (float, optional): the percentage of words in this document
+            which should contain alphabetic character. Defaults to 0.7.
+            Changed from 0.8 in the paper.
+        symbol_2_word_ellipsis (float, optional): The highest acceptable ratio of
+            ellipsis to words. Defaults to 0.1.
+        symbol_2_word_hashtag (float, optional): The highest acceptable ratio of
+            ellipsis to words.. Defaults to 0.1.
+        max_p_begin_bullets (float, optional): Maximum number of lines which begins
+            with a bulletpoint. Defaults to 0.9.
+        max_p_end_ellipsis (float, optional): Maximum number of lines which ends
+            with an ellipsis. Defaults to 0.3.
+    """
+
+    def __init__(
+        self,
+        stop_words: Optional[Set[str]] = None,
+        min_stop_words: int = 2,
+        mean_word_length: Tuple[int, int] = (3, 10),
+        doc_length: Tuple[int, int] = [50, 100_000],
+        alpha_ratio: float = 0.7,
+        symbol_2_word_hashtag: float = 0.1,
+        symbol_2_word_ellipsis: float = 0.1,
+        max_p_begin_bullets: float = 0.9,
+        max_p_end_ellipsis: float = 0.3,
+    ):
+        if stop_words is None:
+            stop_words = set(
+                [
+                    "er",
+                    "jeg",
+                    "det",
+                    "du",
+                    "ikke",
+                    "at",
+                    "en",
+                    "og",
+                    "har",
+                    "vi",
+                    "til",
+                    "på",
+                    "hvad",
+                    "mig",
+                    "med",
+                    "de",
+                    "for",
+                    "den",
+                    "så",
+                    "der",
+                    "dig",
+                    "han",
+                    "kan",
+                    "af",
+                ]
```

Comment:
Burde de her defineres et andet sted og så importes?

Reply:
We should prob. consider deriving this stopword list from somewhere (e.g. 20 most freq. words in DAGW)

Reply:
Jeg submitted danske stopword til spacy for 4 tusinde år siden hvis vi har brug for flere https://github.com/explosion/spaCy/blob/master/spacy/lang/da/stop_words.py

Reply:
Fedt - jeg tænker det er meget godt at have en ranking på hyppigheden af stopwords så man selv kan vælge sige threshold. Kan være jeg ender med at ændre filteret. At bruge spaCy stopwords ville dog generalisere løsningen pænt til andre sprog også.